geo/src/algorithm/relate/geomgraph/intersection_matrix.rs

```diff
@@ -268,6 +278,212 @@ impl IntersectionMatrix {
+        && self.0[CoordPos::Outside][CoordPos::OnBoundary] == Dimensions::Empty ||
+        // [*T****FF*]
+        self.0[CoordPos::Inside][CoordPos::OnBoundary] != Dimensions::Empty
+        && self.0[CoordPos::Outside][CoordPos::Inside] == Dimensions::Empty
+        && self.0[CoordPos::Outside][CoordPos::OnBoundary] == Dimensions::Empty ||
+        // [***T**FF*]
+        self.0[CoordPos::OnBoundary][CoordPos::Inside] != Dimensions::Empty
+        && self.0[CoordPos::Outside][CoordPos::Inside] == Dimensions::Empty
+        && self.0[CoordPos::Outside][CoordPos::OnBoundary] == Dimensions::Empty ||
+        // [****T*FF*]
+        self.0[CoordPos::OnBoundary][CoordPos::OnBoundary] != Dimensions::Empty
+        && self.0[CoordPos::Outside][CoordPos::Inside] == Dimensions::Empty
+        && self.0[CoordPos::Outside][CoordPos::OnBoundary] == Dimensions::Empty
+    }
+
+    /// Returns `true` if `a` touches `b`: they have at least one point in common, but their
+    /// interiors do not intersect.
+    /// # Notes
+    /// - This predicate is **symmetrice**
+    #[allow(clippy::nonminimal_bool)]
+    pub fn is_touches(&self) -> bool {
+        // [FT*******]
+        self.0[CoordPos::Inside][CoordPos::Inside] == Dimensions::Empty
+        && self.0[CoordPos::Outside][CoordPos::Inside] != Dimensions::Empty ||
+        // [F**T*****]
+        self.0[CoordPos::Inside][CoordPos::Inside] == Dimensions::Empty
+        && self.0[CoordPos::OnBoundary][CoordPos::Inside] != Dimensions::Empty ||
+        // [F***T****]
+        self.0[CoordPos::Inside][CoordPos::Inside] == Dimensions::Empty
+        && self.0[CoordPos::OnBoundary][CoordPos::OnBoundary] != Dimensions::Empty
+    }
+
+    /// Compares two geometry objects and returns `true` if their intersection "spatially crosses";
+    /// that is, the geometries have some, but not all interior points in common
+    ///
+    /// ```
+    /// use geo_types::{LineString, line_string, polygon};
+    /// use geo::relate::Relate;
+    ///
+    /// let line_string: LineString = line_string![(x: 85.0, y: 194.0), (x: 162.0, y: 135.0)];
+    /// let poly = polygon![
+    ///     (x: 125., y: 179.),
+    ///     (x: 110., y: 150.),
+    ///     (x: 160., y: 160.),
+    ///     (x: 125., y: 179.),
+    /// ];
+    ///
+    /// let intersection = line_string.relate(&poly);
+    /// assert_eq!(intersection.is_crosses(), true);
+    /// ```
+    /// # Notes
+    /// - If any of the following do not hold, the function will return `false`:
+    ///     - The intersection of the interiors of the geometries must be non-empty
+    ///     - The intersection must have dimension less than the maximum dimension of the two input geometries
+    ///     - The intersection of the two geometries must not equal either geometry
+    /// - This predicate is **symmetric**
+    pub fn is_crosses(&self) -> bool {
+        if self.0[CoordPos::Inside][CoordPos::Outside]
+            != self.0[CoordPos::Outside][CoordPos::Inside]
```

Comment:
There are a few checks in this method that I'm having a hard time reconciling with the logic from the wikipedia page. Did you have another source?

After thinking about it for a while, I think I'm starting to grasp the argument, but I haven't gotten completely through the exercise.

Something like:

```
A's interior cannot be wholly within B's interior unless dims(A.I) <= dims(B.I)
B's interior cannot be wholly within A's interior unless dims(B.I) <= dims(A.I)
```

e.g. a point can never contain a line. a line can never contain a polygon.

So unless `dims(A) == dims(B)` A.I and/or B.I __must__ "leak" into the other's exterior. A

And then we can prove that, when A doesn't contain B, `dims(A.I v B.E) == dims(A.I)` 

Is that about right?

Reply:
the first check: `if self.0[CoordPos::Inside][CoordPos::Outside] != self.0[CoordPos::Outside][CoordPos::Inside]`, establishes that the input geometry dimensions differ, and neither are empty (this works, as demonstrated by the `combination_dimension` test).

If that check fails, they _are_ equal, and we proceed to check whether they're both dimension 1, as that's the only equal dimension geometry that can cross (see the note on https://postgis.net/docs/ST_Crosses.html).

If that first check passes, we proceed to figure out whether dimension A is greater than dimension B, or vice versa, and apply the correct predicate.

Reply:
Are you asserting that `self.0[CoordPos::Inside][CoordPos::Outside] == dims(A)`? That seems like it'd often be true, but I'm not sure what the rules are for when it's true or not. Certainly not true when `B.contains(A)`. What's less clear to me are the boundary conditions.

Is there a way we can make this implementation look more like the definitions on postgis or wikipedia so I wouldn't have to think so hard? 😆

Reply:
My earlier suggestion was something like:
```
let dims_a = max(self.0[CoordPos::Inside][CoordPos::Inside], self.0[CoordPos::Inside][CoordPos::Boundary, self.0[CoordPos::Inside][CoordPos::Exterior]);

let dims_b = max(self.0[CoordPos::Inside][CoordPos::Inside], self.0[CoordPos::Boundary][CoordPos::Inside], self.0[CoordPos::Exterior][CoordPos::Inside]);
```

Then, now that we have `dims_a`/`dims_b`, it seems like we can have a case analysis that aligns with the definitions given in wikipedia and postgis.

If you feel more confident with your approach, I'm ok with going your way, but I'll try to propose some additional comments that cover my gaps in understanding.